### DISCUSSION TITLE
how do I  extend existing heading/paragraphs specs/pluggings to add extra data (like userId)

### DISCUSSION THREAD

**Discussion by rilopez** [bangle-io/bangle.dev#169] [Q&A] [answered]
I am building a  POC  for  collaborative book editor, and I need to keep track of who wrote a given block.  

for example    
if   **alice** type the title `1984` and **bob** type the first paragraph,  the editor will produce this state
```
{
 "type": "doc",
 "content": [
  {
   "type": "heading",
   "attrs": {
    "level": 1,
    "collapseContent": null
   },
   "content": [
    {
     "type": "text",
     "text": "1984"
     "userId":"alice"   <<------ extra attribute,  from the user session 
    }
   ]
  },
  {
   "type": "paragraph",
   "content": [
    {
     "type": "text",
     "text": "It was a bright cold day in April, and the clocks were striking thirteen"
     "userId":"bob"    <<------ extra attribute, from the user session 
    }
   ]
  }
 ]
}
```

I imagine that I have to somehow extend the heading and paragraph plugings/spec.  Could you give some guidance ?

**Comment by kepta** [comment, marked as answer]
It looks like `userId` will be a part of the `attrs`.  `paragraph.spec()` returns a plain object (see [this](https://github.com/kepta/bangle-play/blob/a8e0c38be2abe831fc875cb1980c85c315bb2430/core%2Fcomponents%2Fparagraph.js#L38)) and you can modify that:

```js

export const specRegistry = new SpecRegistry([
... other specs
 {...paragraph.spec(), 
  schema: {
      ...paragraph.spec().schema,
     attrs: { userId: {  default: '' } }
  },
 }
])
```

Read more about node attrs https://prosemirror.net/docs/ref/#model.AttributeSpec

**Reply by rilopez** [reply]
thank you , it works great
